src/modules/submodules/consensus.py

```diff
@@ -82,82 +110,268 @@ def _get_member_info(self, blockstamp: BlockStamp) -> MemberInfo:
+                    'Provided Account is not part of Oracle\'s members and has no submit role. '
                     'For dry mode remove MEMBER_PRIV_KEY from variables.'
                 )
 
         return MemberInfo(
-            is_member=is_member,
-            last_report_ref_slot=last_report_ref_slot,
-            current_ref_slot=current_ref_slot,
-            member_ref_slot=member_ref_slot,
-            member_report_for_current_ref_slot=member_report_for_current_ref_slot,
+            is_report_member=is_member,
+            is_submit_member=is_submit_member,
+            is_fast_line=is_fast_line,
+            fast_line_length_slot=fast_line_length_slot,
+            current_frame_consensus_report=current_frame_consensus_report,
+            current_frame_ref_slot=current_frame_ref_slot,
+            current_frame_member_report=current_frame_member_report,
             deadline_slot=deadline_slot,
         )
 
     @lru_cache(maxsize=1)
-    def _get_current_frame(self, blockstamp: BlockStamp) -> Tuple[SlotNumber, SlotNumber]:
+    def _get_current_frame(self, blockstamp: BlockStamp) -> tuple[SlotNumber, SlotNumber]:
+        """
+        function getCurrentFrame() external view returns (
+            uint256 refSlot,
+            uint256 reportProcessingDeadlineSlot
+        );
+        """
         consensus_contract = self._get_consensus_contract(blockstamp)
         return consensus_contract.functions.getCurrentFrame().call(
             block_identifier=blockstamp.block_hash,
         )
 
+    @lru_cache(maxsize=1)
+    def _get_frame_config(self, blockstamp: BlockStamp) -> tuple[int, int, int]:
+        """
+        struct FrameConfig {
+            uint64 initialEpoch;
+            uint64 epochsPerFrame;
+            uint64 fastLaneLengthSlots;
+        }
+        """
+        consensus_contract = self._get_consensus_contract(blockstamp)
+        return consensus_contract.functions.getFrameConfig().call(
+            block_identifier=blockstamp.block_hash,
+        )
+
+    # ----- Calculation reference slot for report -----
+
     def get_blockstamp_for_report(self, blockstamp: BlockStamp) -> Optional[BlockStamp]:
+        """Get blockstamp that should be used to build and send report for current frame."""
         member_info = self._get_member_info(blockstamp)
 
-        if blockstamp.slot_number < member_info.member_ref_slot:
+        latest_blockstamp = self._get_latest_blockstamp()
+
+        if not self.is_contract_reportable(latest_blockstamp):
+            logger.info({'msg': 'Contract is not reportable.'})
+            return
+
+        if latest_blockstamp.slot_number < member_info.current_frame_ref_slot:
             logger.info({'msg': 'Reference slot is not yet finalized.'})
             return
 
-        # Maybe check head slot number?
-        if blockstamp.slot_number > member_info.deadline_slot:
+        if not member_info.is_fast_line:
```

Comment:
We should decide what word we use - `line` or `lane` as from contracts :)

Reply:
Fixed